compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp

```
@@ -1803,6 +1806,223 @@ std::optional<StringRef> LoweringConfigAttr::getLoweringStrategy() const {
   }
   return std::nullopt;
 }
+constexpr StringLiteral kWorkgroupReorderingStrategyName =
+    "workgroup_reordering_strategy";
+
+::mlir::iree_compiler::IREE::Codegen::WorkgroupReorderingAttrInterface
+LoweringConfigAttr::getWorkgroupReorderingStrategy() const {
+  if (auto attr = getAttributes()
+                      .getAs<::mlir::iree_compiler::IREE::Codegen::
+                                 WorkgroupReorderingAttrInterface>(
+                          kWorkgroupReorderingStrategyName)) {
+    return attr;
+  }
+  return nullptr;
+}
+
+//===----------------------------------------------------------------------===//
+// Conditional Transpose Workgroup Reordering
+//===----------------------------------------------------------------------===//
+
+static std::tuple<SmallVector<OpFoldResult>, SmallVector<OpFoldResult>,
+                  SmallVector<OpFoldResult>>
+getLoopBounds(ArrayRef<Range> loopRanges,
+              ArrayRef<OpFoldResult> givenTileSizes) {
+  SmallVector<OpFoldResult> lbs, ubs, steps;
+  for (auto [loopRange, givenTileSize] :
+       llvm::zip_equal(loopRanges, givenTileSizes)) {
+    // No loop if the tile size is 0.
+    if (isZeroInteger(givenTileSize))
+      continue;
+    lbs.push_back(loopRange.offset);
+    ubs.push_back(loopRange.size);
+    steps.push_back(givenTileSize);
+  }
+  return {lbs, ubs, steps};
+}
+
+static Value computeNumTiles(OpBuilder &b, Location loc, OpFoldResult size,
+                             OpFoldResult tileSize) {
+  Value sizeVal = getValueOrCreateConstantIntOp(b, loc, size);
```

Here and below, I think it might be better to use `affine.apply` directly. So you can do

```
AffineExpr s0, s1;
bindSymbols(b.getContext(), s0, s1);
AffineExpr ceilDivExpr = s0.ceilDiv(s1)
OpFoldResult numTiles = makeComposedFoldedAffineApply(b, loc, ceilDivExpr, {size, tileSize});
```

Also I think this works only with a lower bound of 0, and step of 1. We want to at least make this work with non-zero lower bounds, which case you need to do 
```
affine.apply affine_map<()[s0, s1, s2] -> (s1 - s0) ceilDiv s2>()[%lowerBound, %upperBound, %tileSize]
```

Ideally you make this method take the `step` as well, and assert/error out when  the step is `1`. Today we have no case of `step == 1` (and really no case of `lowerBound != 0` but it could be cause for silent errors).

---

Also, FYI, Prefer to use `divsi` instead of `divui` cause some folding kick in only for `*si`. All of that is managed by the `affine.apply` lowering, so using that will just give you those by default without having to think about it.